pyterrier_colbert/faiss_term_index.py

```diff
@@ -89,33 +89,48 @@ def __init__(self, colbert, index_root, index_name, nprobe=10, partitions=None,
         self.tok = self.inference.query_tokenizer.tok
         vocab_size = self.tok.vocab_size
 
-        if os.path.exists(os.path.join(index_path, 'cfs.np')):
-            self.lookup = np.memmap(os.path.join(index_path, 'cfs.np'), dtype=np.uint64, mode='r')
-        else:
-            print("Computing collection frequencies")
-            self.lookup = torch.zeros(vocab_size, dtype=torch.int64)
-            indx, cnt = self.emb2tid.unique(return_counts=True)
-            self.lookup[indx] += cnt
-            print("Done")
+        if cf:
+            if os.path.exists(os.path.join(index_path, 'cfs.np')):
+                self.lookup = np.memmap(os.path.join(index_path, 'cfs.np'), dtype=np.uint64, mode='r')
+            else:
+                samplerange=len(self.emb2tid)
+                samplerange = int(samplerange*stats_sample_fraction)
+                print("Computing collection frequencies")
+                #self.lookup = torch.zeros(vocab_size, dtype=torch.int64)
+                self.lookup = np.zeros(vocab_size, dtype=np.int64)
+                #.astype(np.int64, copy=False)
+                indx, cnt = np.unique(self.emb2tid[0:samplerange], return_counts=True)
+                self.lookup[indx] += cnt
+                print("Done")
         
         if os.path.exists(os.path.join(index_path, 'doclens.psum.np')):
             self.doclens = None
             self.end_offsets = np.memmap(os.path.join(index_path, 'doclens.psum.np'), dtype=np.uint64, mode='r')
             self.num_docs = self.end_offsets.shape[0] - 1
+            #inverse of cumsum
+            #https://stackoverflow.com/questions/38666924/what-is-the-inverse-of-the-numpy-cumsum-function
+            self.doclens = self.end_offsets.copy()            
```

Comment:
We need doclens to compute the DFs. Yes, it would be better to compute it on the fly. But is doclens really that large that it cant ever be loaded into memory?

Reply:
i.e. 
```
for docid in tqdm(selfrange(numdocs)):
  doclen = self.end_offsets[docid] - self.end_offsets[docid-1] etc
```

Reply:
BTW, I found end_offsets had 0 as the first value, which surprised me, as I know the first passage in the index isnt empty.

Reply:
> We need doclens to compute the DFs

DFs are just pre-computed when the number of docs is unmanageable, though.

> BTW, I found end_offsets had 0

Yeah, so the file contains one more element than there are documents. It's meant to be used as `start_idx, stop_idx = self.doc_offsets[docid:docid+2]`. This way, it can correctly work without additional code for docid 0. (The above sample loops back around to the last document and gives a very large negative number.)

Reply:
If we choose to continue the doclen compute-on-the-fly approach, we have some fixes to make.

Reply:
Or, alternatively, we can just use the `doclens.np` (no psum) file.

Reply:
go for it.